src/lib/actor/actor.ml

```diff
@@ -0,0 +1,260 @@
+
+  type process_status = Status_processed | Status_fallthrough | Status_exit
+
+  let spawn (actor : _ t) : unit Deferred.Or_error.t =
+    let process_msg ~state ~deque ~handler () =
+      match Deque.dequeue_front deque with
+      | Some message -> (
+          match%map handler ~state ~message with
+          | MUnprocessed ->
+              Deque.enqueue_front deque message ;
+              MUnprocessed
+          | result ->
+              result )
+      | None ->
+          Deferred.return MUnprocessed
+    in
+    let process_request ~state ~(deque : request_e Deque.t)
+        ~(handler : _ request_handler) () =
+      match Deque.dequeue_front deque with
+      | Some (EReq (request, response_ivar) as e) -> (
+          match%map handler.f ~state ~request with
+          | RNext (state, resp) ->
+              Ivar.fill response_ivar resp ;
+              MNext state
+          | RExit resp ->
+              (* NOTE: for some reason OCaml can't infer this type here *)
+              Ivar.fill response_ivar (Obj.magic resp) ;
+              MExit
+          | RUnprocessed ->
+              Deque.enqueue_front deque e ;
+              MUnprocessed )
+      | None ->
+          Deferred.return MUnprocessed
+    in
+
+    if actor.is_running then
+      Deferred.Or_error.of_exn (RunningActorSpawned { name = actor.name })
+    else (
+      actor.is_running <- true ;
+      let rec loop () =
+        if not actor.is_running then Deferred.unit
+        else
+          let dispatchers =
+            [ process_request ~state:actor.state ~deque:actor.request_inbox
+                ~handler:actor.request_handler
+            ; process_msg ~state:actor.state ~deque:actor.control_inbox
+                ~handler:actor.control_handler
+            ; process_msg ~state:actor.state ~deque:actor.data_inbox
+                ~handler:actor.data_handler
+            ]
+          in
+          let rec iter_dispatch = function
+            | [] ->
+                Deferred.return MUnprocessed
+            | dispatcher :: dispatchers -> (
+                match%bind dispatcher () with
+                | MUnprocessed ->
+                    iter_dispatch dispatchers
+                | (MExit | MNext _) as processed ->
+                    Deferred.return processed )
+          in
+          match%bind iter_dispatch dispatchers with
+          | MNext state ->
+              actor.state <- state ;
+              loop ()
+          | MExit ->
+              actor.is_running <- false ;
+              Deferred.unit
+          | MUnprocessed ->
+              let%bind () = Async.Scheduler.yield () in
+              loop ()
+      in
+
+      let%bind.Deferred () = loop () in
+      Deferred.Or_error.return () )
+end
+
+module Regular (DataMessage : sig
+  type t
+
+  val to_yojson : t -> Yojson.Safe.t
+end) =
+struct
+  module Inner = WithRequest (DataMessage) (DummyRequest)
+  include Inner
+
+  let create = create ~request_handler:{ f = DummyRequest.handler }
```

Comment:
Should shadow the function `send_request here`, I guess could go into interfaces?